prototype2/client/infra/dag/src/lib.rs

```diff
@@ -0,0 +1,242 @@
+  /// should go first
+  pub fn add_dependency(&mut self, system_name: &I, other: &(I, String)) {
+    let &(ref other_name, ref other_alias) = other;
+    if system_name == other_name {
+      println!("Tried to make {:?} depend on itself!", system_name);
+      panic!("Dag dependency error! Tried to depend on self! Check stdout for details.");
+    }
+    {
+      let mut name_set = self.name_to_fancy_names.entry(other_name.clone()).or_insert(Vec::new());
+      name_set.push(other_alias.clone());
+    }
+
+    self.add_system(system_name);
+    self.add_system(other_name);
+
+    let system_node = self.name_to_idx.get(system_name).unwrap();
+    let other_node = self.name_to_idx.get(other_name).unwrap();
+
+    match self.internal_dag.add_edge(system_node.clone(), other_node.clone(), ()) {
+      Err(daggy::WouldCycle(_)) => self.cycle_panic(system_name, other_name),
+      _ => (),
+    }
+  }
+
+  /// Throw a whole bunch of deps in at once
+  pub fn add_dependency_set(&mut self, system_name: &I, others: &[(I, String)]) {
+    others.iter().foreach(|other| {
+      self.add_dependency(system_name, other);
+    })
+  }
+
+  /// Internal method that panics if the system already exists
+  pub fn add_system(&mut self, system_name: &I) {
+    if !self.name_to_idx.contains_key(system_name) {
+      self.force_add_system(system_name);
+    }
+  }
+
+  /// Internal method that panics if the system already exists
+  fn force_add_system(&mut self, system_name: &I) {
+    assert!(!self.name_to_idx.contains_key(system_name),
+            "Internal add_system method was called with system that already exists");
+
+    let node = self.internal_dag.add_node(());
+    self.name_to_idx.insert(system_name.clone(), node);
+  }
+
+  /// Diagnostic method for panicking on edge insert and printing the dag that
+  /// caused the panic
+  fn cycle_panic(&self, system_name: &I, other_name: &I) {
+    println!("Dag error on add_dependency! Tried to add a dependency from ({:?}) on ({:?})",
+             system_name,
+             other_name);
+    println!("The existing dag was:");
+    self.print();
+
+    panic!("Dag cycle error! Check stdout for the offending nodes and dag");
+  }
+
+  /// Print the dag with the system names
+  pub fn print(&self) {
+    let name_to_idx = &self.name_to_idx;
+    let internal_dag = &self.internal_dag;
+    let mut idx_to_name = HashMap::new();
+    name_to_idx.iter().foreach(|(k, v)| {
+      idx_to_name.insert(v, k);
+    });
+    let mut name_to_fancy_name = self.simplify_fancy_names();
+
+    println!("");
+    name_to_idx.iter().foreach(|(k, v)| {
+      println!("{}: [", fancify_name(k, &name_to_fancy_name));
+      internal_dag.children(v.clone())
+        .iter(&internal_dag)
+        .map(|(_, n)| idx_to_name.get(&n).unwrap())
+        .foreach(|name| println!("  {}", fancify_name(name.clone(), &name_to_fancy_name)));
+      println!("]");
+    });
+  }
+
+  fn simplify_fancy_names(&self) -> HashMap<I, String> {
```

Comment:
I added this instrumentation for the dep reflection really late. Thats why the methods all have terrible names (*fancy*), and the string just looks tacked on everywhere. Thats because it is.

I think i'll add another method to "DependencyAware" and use that to identify dependencies (that is, convert from TypeId to a human readable string).. Then we don't have to do this nonsense.